lighthouse-treemap/app/src/main.js

```diff
@@ -9,6 +9,244 @@
 
 /* globals webtreemap TreemapUtil */
 
+/** @type {TreemapViewer} */
+let treemapViewer;
+
+class TreemapViewer {
+  /**
+   * @param {LH.Treemap.Options} options
+   * @param {HTMLElement} el
+   */
+  constructor(options, el) {
+    const treemapDebugData = /** @type {LH.Audit.Details.DebugData} */ (
+      options.lhr.audits['script-treemap-data'].details);
+    if (!treemapDebugData || !treemapDebugData.treemapData) {
+      throw new Error('missing script-treemap-data');
+    }
+
+    /** @type {import('../../../lighthouse-core/audits/script-treemap-data').TreemapData} */
+    const scriptRootNodes = treemapDebugData.treemapData;
+
+    /** @type {WeakMap<LH.Treemap.Node, LH.Treemap.RootNodeContainer>} */
+    this.nodeToRootNodeMap = new WeakMap();
+
+    /** @type {{[group: string]: LH.Treemap.RootNodeContainer[]}} */
+    this.rootNodesByGroup = {
+      scripts: scriptRootNodes,
+    };
+
+    for (const rootNodes of Object.values(this.rootNodesByGroup)) {
+      for (const rootNode of rootNodes) {
+        TreemapUtil.dfs(rootNode.node, node => this.nodeToRootNodeMap.set(node, rootNode));
+      }
+    }
+
+    /** @type {LH.Treemap.Node} */
+    this.currentRootNode; // eslint-disable-line no-unused-expressions
+    this.documentUrl = options.lhr.requestedUrl;
+    this.el = el;
+    this.getHue = TreemapUtil.stableHasher(TreemapUtil.COLOR_HUES);
+
+    this.createHeader();
+    this.show();
+    this.initListeners();
+  }
+
+  createHeader() {
+    TreemapUtil.find('.lh-header--url').textContent = this.documentUrl;
+    TreemapUtil.find('.lh-header--size').textContent =
+      TreemapUtil.formatBytes(this.createRootNodeForGroup('scripts').resourceBytes);
+  }
+
+  initListeners() {
+    window.addEventListener('resize', () => {
+      this.render();
```

Comment:
one of the changes i made is to expose a `webtreemap.layout` method that'll relayout an existing treemap, instead of rebuilding the entire DOM again.

so i'd recommend doing that for a lil perf win.

(it was substantial with bigass trees)

This also means you wont have to `updateColors` on a resize.

Reply:
done, but still need to update colors, newly added nodes (b/c of bigger viewport) need to be colored too.